File: Tanks.java
// Tanks.java
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.util.*;

class Bullet {
    int x, y, dx, dy, bounces;
    boolean alive;
    Bullet(int x, int y, int dx, int dy, int bounces) {
        this.x=x; this.y=y; this.dx=dx; this.dy=dy; this.bounces=bounces; alive=true;
    }
}

public class Tanks extends JPanel implements ActionListener, KeyListener {
    static final int W = 40, H = 20, CELL = 20;
    int pX = W/2, pY = H/2, pDX = 0, pDY = -1;
    char pSym = '▲';
    java.util.List<int[]> enemies = new ArrayList<>();
    java.util.List<Bullet> bullets = new ArrayList<>();
    int score = 0, lives = 3;
    boolean gameOver = false;
    Timer timer;
    Random rand = new Random();
    int spawnTimer = 0, enemySpawnInterval = 40;
    int shootCooldown = 0, shootDelay = 15;

    public Tanks() {
        setPreferredSize(new Dimension(W*CELL + 150, H*CELL));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);
        timer = new Timer(50, this);
        timer.start();
    }

    void spawnEnemy() {
        int side = rand.nextInt(4);
        int x=0, y=0;
        switch(side) {
            case 0: x = rand.nextInt(W-4)+2; y = 2; break;
            case 1: x = rand.nextInt(W-4)+2; y = H-3; break;
            case 2: x = 2; y = rand.nextInt(H-4)+2; break;
            default: x = W-3; y = rand.nextInt(H-4)+2;
        }
        if (Math.abs(x-pX) < 3 && Math.abs(y-pY) < 3) return;
        int[] dirs = {0,1,0,-1,1,0,-1,0};
        int d = rand.nextInt(4)*2;
        enemies.add(new int[]{x, y, dirs[d], dirs[d+1], '▼', rand.nextInt(60)+30});
    }

    void shoot() {
        if (shootCooldown > 0) return;
        bullets.add(new Bullet(pX+pDX, pY+pDY, pDX, pDY, 3));
        shootCooldown = shootDelay;
    }

    void enemyShoot(int[] e) {
        bullets.add(new Bullet(e[0]+e[2], e[1]+e[3], e[2], e[3], 2));
    }

    void update() {
        if (gameOver) return;
        if (shootCooldown > 0) shootCooldown--;
        spawnTimer++;
        if (spawnTimer >= enemySpawnInterval && enemies.size() < 6) {
            spawnEnemy();
            spawnTimer = 0;
        }
        for (int i = 0; i < enemies.size(); i++) {
            int[] e = enemies.get(i);
            int dx = pX - e[0], dy = pY - e[1];
            if (Math.abs(dx) > Math.abs(dy)) e[0] += dx > 0 ? 1 : -1;
            else e[1] += dy > 0 ? 1 : -1;
            e[0] = Math.max(2, Math.min(W-3, e[0]));
            e[1] = Math.max(2, Math.min(H-3, e[1]));
            if (Math.abs(dx) > Math.abs(dy)) { e[2] = dx > 0 ? 1 : -1; e[3] = 0; }
            else { e[2] = 0; e[3] = dy > 0 ? 1 : -1; }
            e[5]--;
            if (e[5] <= 0 && rand.nextDouble() < 0.3) {
                enemyShoot(e);
                e[5] = rand.nextInt(80)+40;
            }
            if (e[0] == pX && e[1] == pY) {
                lives--;
                if (lives <= 0) gameOver = true;
                else { pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲'; }
                enemies.set(i, e);
                return;
            }
            enemies.set(i, e);
        }
        for (int i = 0; i < bullets.size(); i++) {
            Bullet b = bullets.get(i);
            if (!b.alive) { bullets.remove(i); i--; continue; }
            int nx = b.x + b.dx, ny = b.y + b.dy;
            boolean bounced = false;
            if (nx <= 1 || nx >= W-2) { b.dx *= -1; b.bounces--; bounced = true; }
            if (ny <= 1 || ny >= H-2) { b.dy *= -1; b.bounces--; bounced = true; }
            if (bounced) { b.x += b.dx; b.y += b.dy; }
            else { b.x = nx; b.y = ny; }
            if (b.bounces <= 0 || b.x <= 0 || b.x >= W-1 || b.y <= 0 || b.y >= H-1) {
                b.alive = false;
                bullets.remove(i); i--;
                continue;
            }
            boolean hit = false;
            for (int j = 0; j < enemies.size(); j++) {
                int[] e = enemies.get(j);
                if (e[0] == b.x && e[1] == b.y) {
                    enemies.remove(j);
                    bullets.remove(i);
                    score += 10;
                    hit = true;
                    i--;
                    break;
                }
            }
            if (hit) continue;
            if (b.x == pX && b.y == pY) {
                lives--;
                bullets.remove(i);
                i--;
                if (lives <= 0) gameOver = true;
                else { pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲'; }
            }
        }
    }

    void movePlayer(int dx, int dy) {
        if (gameOver) return;
        int nx = pX + dx, ny = pY + dy;
        if (nx >= 2 && nx < W-2 && ny >= 2 && ny < H-2) {
            pX = nx; pY = ny;
            if (dx != 0 || dy != 0) {
                pDX = dx; pDY = dy;
                if (dx == 1) pSym = '►';
                else if (dx == -1) pSym = '◄';
                else if (dy == -1) pSym = '▲';
                else if (dy == 1) pSym = '▼';
            }
        }
    }

    void restart() {
        pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲';
        enemies.clear(); bullets.clear();
        score = 0; lives = 3; gameOver = false; spawnTimer = 0;
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        update();
        repaint();
    }

    @Override
    public void paintComponent(Graphics g) {
        super.paintComponent(g);
        g.setColor(Color.WHITE);
        // border
        for (int x = 0; x < W; x++) {
            g.drawRect(x*CELL, 0, CELL, CELL);
            g.drawRect(x*CELL, (H-1)*CELL, CELL, CELL);
        }
        for (int y = 0; y < H; y++) {
            g.drawRect(0, y*CELL, CELL, CELL);
            g.drawRect((W-1)*CELL, y*CELL, CELL, CELL);
        }
        // player
        g.setColor(Color.GREEN);
        g.fillRect(pX*CELL, pY*CELL, CELL, CELL);
        g.setColor(Color.BLACK);
        g.drawString(String.valueOf(pSym), pX*CELL+5, pY*CELL+15);
        // enemies
        g.setColor(Color.RED);
        for (int[] e : enemies) {
            g.fillRect(e[0]*CELL, e[1]*CELL, CELL, CELL);
            g.setColor(Color.BLACK);
            g.drawString("E", e[0]*CELL+5, e[1]*CELL+15);
            g.setColor(Color.RED);
        }
        // bullets
        g.setColor(Color.YELLOW);
        for (Bullet b : bullets) {
            if (b.alive) g.fillOval(b.x*CELL+5, b.y*CELL+5, 10, 10);
        }
        // info
        g.setColor(Color.WHITE);
        g.drawString("Score: "+score, W*CELL+10, 20);
        g.drawString("Lives: "+lives, W*CELL+10, 40);
        g.drawString("Enemies: "+enemies.size(), W*CELL+10, 60);
        g.drawString("WASD move, Space fire", W*CELL+10, 80);
        if (gameOver) {
            g.setColor(Color.RED);
            g.setFont(new Font("Arial", Font.BOLD, 24));
            g.drawString("GAME OVER", W*CELL/2-50, H*CELL/2);
            g.setFont(new Font("Arial", Font.PLAIN, 14));
            g.drawString("Press R to restart", W*CELL/2-60, H*CELL/2+30);
        }
    }

    @Override
    public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        if (key == KeyEvent.VK_R && gameOver) restart();
        if (key == KeyEvent.VK_SPACE) shoot();
        if (key == KeyEvent.VK_W || key == KeyEvent.VK_UP) movePlayer(0, -1);
        if (key == KeyEvent.VK_S || key == KeyEvent.VK_DOWN) movePlayer(0, 1);
        if (key == KeyEvent.VK_A || key == KeyEvent.VK_LEFT) movePlayer(-1, 0);
        if (key == KeyEvent.VK_D || key == KeyEvent.VK_RIGHT) movePlayer(1, 0);
        if (key == KeyEvent.VK_Q) System.exit(0);
    }
    @Override public void keyReleased(KeyEvent e) {}
    @Override public void keyTyped(KeyEvent e) {}

    public static void main(String[] args) {
        JFrame frame = new JFrame("Tanks (Ricochet)");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setResizable(false);
        frame.add(new Tanks());
        frame.pack();
        frame.setLocationRelativeTo(null);
        frame.setVisible(true);
    }
}

File: Tanks.cs
// Tanks.cs
using System;
using System.Collections.Generic;
using System.Threading;

class Bullet
{
    public int X, Y, DX, DY, Bounces;
    public bool Alive;
    public Bullet(int x, int y, int dx, int dy, int bounces = 3)
    {
        X = x; Y = y; DX = dx; DY = dy; Bounces = bounces; Alive = true;
    }
    public void Move(int w, int h)
    {
        int nx = X + DX, ny = Y + DY;
        bool bounced = false;
        if (nx <= 1 || nx >= w-2) { DX *= -1; Bounces--; bounced = true; }
        if (ny <= 1 || ny >= h-2) { DY *= -1; Bounces--; bounced = true; }
        if (bounced) { X += DX; Y += DY; }
        else { X = nx; Y = ny; }
        if (Bounces <= 0 || X <= 0 || X >= w-1 || Y <= 0 || Y >= h-1)
            Alive = false;
    }
}

class Tanks
{
    const int W = 40, H = 20;
    static Random rand = new Random();
    static int pX, pY, pDX, pDY;
    static char pSym;
    static List<(int x, int y, int dx, int dy, char sym, int shootTimer)> enemies =
        new List<(int, int, int, int, char, int)>();
    static List<Bullet> bullets = new List<Bullet>();
    static int score = 0, lives = 3;
    static bool gameOver = false, running = true;
    static int spawnTimer = 0, enemySpawnInterval = 40;
    static int shootCooldown = 0, shootDelay = 15;

    static void DrawBorder()
    {
        for (int x = 0; x < W; x++)
        {
            Console.SetCursorPosition(x, 0); Console.Write('#');
            Console.SetCursorPosition(x, H-1); Console.Write('#');
        }
        for (int y = 0; y < H; y++)
        {
            Console.SetCursorPosition(0, y); Console.Write('#');
            Console.SetCursorPosition(W-1, y); Console.Write('#');
        }
    }

    static void SpawnEnemy()
    {
        int side = rand.Next(4);
        int x=0, y=0;
        switch(side)
        {
            case 0: x = rand.Next(2, W-2); y = 2; break;
            case 1: x = rand.Next(2, W-2); y = H-3; break;
            case 2: x = 2; y = rand.Next(2, H-2); break;
            default: x = W-3; y = rand.Next(2, H-2); break;
        }
        if (Math.Abs(x-pX) < 3 && Math.Abs(y-pY) < 3) return;
        int[] dirs = {0,1,0,-1,1,0,-1,0};
        int d = rand.Next(4)*2;
        enemies.Add((x, y, dirs[d], dirs[d+1], '▼', rand.Next(30, 90)));
    }

    static void Shoot()
    {
        if (shootCooldown > 0) return;
        bullets.Add(new Bullet(pX + pDX, pY + pDY, pDX, pDY, 3));
        shootCooldown = shootDelay;
    }

    static void EnemyShoot(ref (int x, int y, int dx, int dy, char sym, int shootTimer) e)
    {
        bullets.Add(new Bullet(e.x + e.dx, e.y + e.dy, e.dx, e.dy, 2));
    }

    static void Update()
    {
        if (gameOver) return;
        if (shootCooldown > 0) shootCooldown--;
        spawnTimer++;
        if (spawnTimer >= enemySpawnInterval && enemies.Count < 6)
        {
            SpawnEnemy();
            spawnTimer = 0;
        }
        for (int i = 0; i < enemies.Count; i++)
        {
            var e = enemies[i];
            int dx = pX - e.x, dy = pY - e.y;
            if (Math.Abs(dx) > Math.Abs(dy))
                e.x += dx > 0 ? 1 : -1;
            else
                e.y += dy > 0 ? 1 : -1;
            e.x = Math.Max(2, Math.Min(W-3, e.x));
            e.y = Math.Max(2, Math.Min(H-3, e.y));
            if (Math.Abs(dx) > Math.Abs(dy))
                { e.dx = dx > 0 ? 1 : -1; e.dy = 0; }
            else
                { e.dx = 0; e.dy = dy > 0 ? 1 : -1; }
            e.shootTimer--;
            if (e.shootTimer <= 0 && rand.NextDouble() < 0.3)
            {
                EnemyShoot(ref e);
                e.shootTimer = rand.Next(40, 120);
            }
            if (e.x == pX && e.y == pY)
            {
                lives--;
                if (lives <= 0) gameOver = true;
                else { pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲'; }
                enemies[i] = e;
                return;
            }
            enemies[i] = e;
        }
        for (int i = 0; i < bullets.Count; i++)
        {
            var b = bullets[i];
            if (!b.Alive) { bullets.RemoveAt(i); i--; continue; }
            b.Move(W, H);
            if (!b.Alive) { bullets.RemoveAt(i); i--; continue; }
            bool hit = false;
            for (int j = 0; j < enemies.Count; j++)
            {
                if (enemies[j].x == b.X && enemies[j].y == b.Y)
                {
                    enemies.RemoveAt(j);
                    bullets.RemoveAt(i);
                    score += 10;
                    hit = true;
                    i--;
                    break;
                }
            }
            if (hit) continue;
            if (b.X == pX && b.Y == pY)
            {
                lives--;
                bullets.RemoveAt(i);
                i--;
                if (lives <= 0) gameOver = true;
                else { pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲'; }
            }
        }
    }

    static void MovePlayer(int dx, int dy)
    {
        if (gameOver) return;
        int nx = pX + dx, ny = pY + dy;
        if (nx >= 2 && nx < W-2 && ny >= 2 && ny < H-2)
        {
            pX = nx; pY = ny;
            if (dx != 0 || dy != 0)
            {
                pDX = dx; pDY = dy;
                if (dx == 1) pSym = '►';
                else if (dx == -1) pSym = '◄';
                else if (dy == -1) pSym = '▲';
                else if (dy == 1) pSym = '▼';
            }
        }
    }

    static void Render()
    {
        Console.Clear();
        DrawBorder();
        Console.SetCursorPosition(pX, pY); Console.Write(pSym);
        foreach (var e in enemies)
        {
            Console.SetCursorPosition(e.x, e.y);
            Console.Write(e.sym);
        }
        foreach (var b in bullets)
            if (b.Alive)
            {
                Console.SetCursorPosition(b.X, b.Y);
                Console.Write('*');
            }
        Console.SetCursorPosition(W+2, 0);
        Console.Write($"Score: {score}  Lives: {lives}");
        Console.SetCursorPosition(W+2, 1);
        Console.Write($"Enemies: {enemies.Count}");
        Console.SetCursorPosition(W+2, 2);
        Console.Write("WASD move, Space fire");
        if (gameOver)
        {
            Console.SetCursorPosition(W/2-5, H/2);
            Console.Write("GAME OVER! Press R");
        }
    }

    static void Restart()
    {
        pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲';
        enemies.Clear();
        bullets.Clear();
        score = 0; lives = 3; gameOver = false; spawnTimer = 0;
    }

    static void InputLoop()
    {
        while (running)
        {
            var key = Console.ReadKey(true);
            switch (key.Key)
            {
                case ConsoleKey.R:
                    if (gameOver) Restart();
                    break;
                case ConsoleKey.Spacebar:
                    Shoot();
                    break;
                case ConsoleKey.W:
                case ConsoleKey.UpArrow:
                    MovePlayer(0, -1);
                    break;
                case ConsoleKey.S:
                case ConsoleKey.DownArrow:
                    MovePlayer(0, 1);
                    break;
                case ConsoleKey.A:
                case ConsoleKey.LeftArrow:
                    MovePlayer(-1, 0);
                    break;
                case ConsoleKey.D:
                case ConsoleKey.RightArrow:
                    MovePlayer(1, 0);
                    break;
                case ConsoleKey.Q:
                    running = false;
                    return;
            }
        }
    }

    static void Main()
    {
        Console.CursorVisible = false;
        pX = W/2; pY = H/2; pDX = 0; pDY = -1; pSym = '▲';
        Thread inputThread = new Thread(InputLoop);
        inputThread.IsBackground = true;
        inputThread.Start();
        while (running)
        {
            Update();
            Render();
            Thread.Sleep(50);
        }
    }
}
